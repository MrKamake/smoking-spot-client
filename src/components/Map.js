import React, { useState, useEffect } from 'react';
import './Map.scss';
import { Link } from 'react-router-dom';
import {
  withGoogleMap,
  GoogleMap,
  Marker,
  InfoWindow
} from 'react-google-maps';
import { MarkerClusterer } from 'react-google-maps/lib/components/addons/MarkerClusterer';
import PlacesAutocomplete, {
  geocodeByAddress,
  getLatLng
} from 'react-places-autocomplete';

const Map = ({ getSpotList, spotList }) => {
  const [location, setLocation] = useState({});
  const [searchBar, setSearchBar] = useState(false);
  const [address, setAddress] = useState('');
  const [isOpen, setIsOpen] = useState(false);

  useEffect(() => {
    getGeolocation();
    getSpotList();
    return () => {
      console.log('clean up!');
    };
  }, []);

  const RenderMap = withGoogleMap(props => (
    <GoogleMap
      defaultCenter={{ lat: location.lat, lng: location.lng }}
      defaultZoom={17}
    >
      {spotList.length ? (
        <MarkerClusterer averageCenter enableRetinaIcons gridSize={60}>
          {spotList.map((spot, index) => (
            <Marker
              key={index}
              position={{
                lng: spot.location.coordinates[0],
                lat: spot.location.coordinates[1]
              }}
              onClick={() => {
                setLocation({
                  lng: spot.location.coordinates[0],
                  lat: spot.location.coordinates[1]
                });
                setIsOpen(true);
              }}
            >
              {isOpen && (
                <InfoWindow
                  className='info-window'
                  onCloseClick={() => setIsOpen(false)}
                >
                  <Link
                    to={`/spots/${spot._id}`}
                    className='spot-description-box'
                  >
                    <img
                      className='spot-image'
                      src={spot.photo_url}
                      alt={spot.title}
                    />
                    <div className='spot-title'>{spot.title}</div>
                    <div className='spot-address'>{spot.address}</div>
                  </Link>
                </InfoWindow>
              )}
            </Marker>
          ))}
        </MarkerClusterer>
      ) : null}
    </GoogleMap>
  ));

  const getGeolocation = () => {
    const geo = navigator.geolocation;
    geo.getCurrentPosition(({ coords }) => {
      setLocation({
        lat: coords.latitude,
        lng: coords.longitude
      });
    });
  };

  const _handleClick = boolean => {
    setSearchBar(boolean);
  };

  const _handleSelect = addressData => {
    geocodeByAddress(addressData)
      .then(result => getLatLng(result[0]))
      .then(latLng => setLocation(latLng))
      .catch(err => console.log('error', err));
  };

  return (
    <>
      {console.log('spotList', spotList)}
      {searchBar ? (
        <div className='search-bar'>
          <img
            className='search-button-icon'
            alt='search bar'
            src='https://www.iconsdb.com/icons/preview/white/search-12-xxl.png'
            onClick={() => _handleClick(false)}
          />
          <PlacesAutocomplete
            value={address}
            onChange={addressData => setAddress(addressData)}
            onSelect={_handleSelect}
          >
            {({
              getInputProps,
              suggestions,
              getSuggestionItemProps,
              loading
            }) => (
              <>
                <input
                  {...getInputProps({
                    placeholder: 'Search Places ...',
                    className: 'location-search-input'
                  })}
                />
                <div className='autocomplete-dropdown-container'>
                  {loading && <div>Loading...</div>}
                  {suggestions.map(suggestion => {
                    const className = suggestion.active
                      ? 'suggestion-item--active'
                      : 'suggestion-item';
                    const style = suggestion.active
                      ? { backgroundColor: '#fafafa', cursor: 'pointer' }
                      : { backgroundColor: '#ffffff', cursor: 'pointer' };
                    return (
                      <div
                        {...getSuggestionItemProps(suggestion, {
                          className,
                          style
                        })}
                      >
                        <span>{suggestion.description}</span>
                      </div>
                    );
                  })}
                </div>
              </>
            )}
          </PlacesAutocomplete>
        </div>
      ) : (
        <div className='header-box'>
          <Link to='/spots/new' className='new-page-button'>
            <img
              className='new-button-icon'
              alt='new-button'
              src='https://www.iconsdb.com/icons/preview/white/plus-5-xxl.png'
            />
          </Link>
          <Link to='/home' className='sub-logo'>
            <div className='sub-logo-smoking'>SMOKING</div>
            <div className='sub-logo-spot'>SPOT</div>
          </Link>
          <button className='search-button' onClick={() => _handleClick(true)}>
            <img
              className='search-button-icon'
              alt='search box'
              src='https://www.iconsdb.com/icons/preview/white/search-12-xxl.png'
            />
          </button>
        </div>
      )}
      <div className='map-box'>
        <RenderMap
          containerElement={<div style={{ height: `715px`, width: '375px' }} />}
          mapElement={<div style={{ height: `100%` }} />}
        />
      </div>
    </>
  );
};

export default Map;
