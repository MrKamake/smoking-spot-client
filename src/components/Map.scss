.search-bar {
  position: relative;
  padding: 30px 0 20px;
  .search-button-icon {
    width: 30px;
    padding-right: 15px;
  }
}
.gm-style img {
  max-width: 200px;
}
a:-webkit-any-link {
  text-decoration: none;
  color: black;
  .spot-title {
    font-size: 1.3rem;
    font-weight: bold;
  }
  .spot-address {
    font-weight: bold;
    color: darkgray;
  }
}
.autocomplete-dropdown-container {
  position: absolute;
  z-index: 9999;
}
.location-search-input {
  width: 300px;
  height: 40px;
  line-height: 50px;
  border: 1px solid lightgrey;
  text-decoration: none;
  font-size: 1rem;
  border-radius: 5px;
}
