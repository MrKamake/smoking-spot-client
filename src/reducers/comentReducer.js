import { initState } from './index';
import { GET_COMENTS } from '../constants/actionTypes';

export const comentReducer = (state = initState.coments, action) => {
  switch (action.type) {
    case GET_COMENTS:
      return action.coments;
    default:
      return state;
  }
};
