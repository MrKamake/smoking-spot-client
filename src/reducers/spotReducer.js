import { initState } from './index';
import { GET_SPOT_LIST, GET_SPOT_DETAIL } from '../constants/actionTypes';

export const spotReducer = (state = initState.spotList, action) => {
  switch (action.type) {
    case GET_SPOT_LIST:
      return action.spotList;
    case GET_SPOT_DETAIL:
      return action.spot;
    default:
      return state;
  }
};
