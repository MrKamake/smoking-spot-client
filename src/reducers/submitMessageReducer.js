import { initState } from './index';
import { STORE_MESSAGE } from '../constants/actionTypes';

export const submitMessageReducer = (
  state = initState.submitMessage,
  action
) => {
  switch (action.type) {
    case STORE_MESSAGE:
      return action.msg;
    default:
      return state;
  }
};
