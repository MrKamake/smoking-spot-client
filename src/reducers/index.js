import { combineReducers } from 'redux';
import { spotReducer } from './spotReducer';
import { comentReducer } from './comentReducer';
import { submitMessageReducer } from './submitMessageReducer';

export const initState = {
  spotList: [],
  coments: [],
  submitMessage: ''
};

export default combineReducers({
  spotReducer,
  comentReducer,
  submitMessageReducer
});
