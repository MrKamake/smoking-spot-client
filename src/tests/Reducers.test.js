import { initState } from '../reducers/index';

import { spotReducer } from '../reducers/spotReducer';
import { comentReducer } from '../reducers/comentReducer';
import { submitMessageReducer } from '../reducers/submitMessageReducer';

import { spotList } from '../tests/Map.test';
import {
  GET_SPOT_LIST,
  GET_SPOT_DETAIL,
  GET_COMENTS,
  STORE_MESSAGE
} from '../constants/actionTypes';

describe('reducers', () => {
  it('should return init value', () => {
    expect(initState).toHaveProperty('spotList');
    expect(initState).toHaveProperty('coments');
    expect(initState.submitMessage).toEqual('');
    expect(initState.spotList).toEqual([]);
    expect(spotReducer(undefined, {})).toEqual(initState.spotList);
    expect(comentReducer(undefined, {})).toEqual(initState.coments);
    expect(submitMessageReducer(undefined, {})).toEqual(
      initState.submitMessage
    );
  });
  describe('data of action should return equal value', () => {
    const coments = ['hahaha', 'smile', 'kiki'];
    it('in spotReducer', () => {
      expect(
        spotReducer(initState.spotList, {
          type: GET_SPOT_LIST,
          spotList
        })
      ).toEqual(spotList);
      expect(
        spotReducer(spotList, {
          type: GET_SPOT_DETAIL,
          spot: 'seoul'
        })
      ).toEqual('seoul');
    });

    it('in comentReducer', () => {
      expect(
        comentReducer(initState.coments, {
          type: GET_COMENTS,
          coments
        })
      ).toEqual([...coments]);
      expect(
        comentReducer([...coments], {
          type: GET_COMENTS,
          coments: [...coments, 'hello', 'hi']
        })
      ).toEqual([...coments, 'hello', 'hi']);
    });

    it('in submitMessageReducer', () => {
      expect(
        submitMessageReducer(initState.submitMessage, {
          type: STORE_MESSAGE,
          msg: 'Success!!'
        })
      ).toEqual('Success!!');
      expect(
        submitMessageReducer(initState.submitMessage, {
          type: STORE_MESSAGE,
          msg: 'Fail..'
        })
      ).toEqual('Fail..');
    });
  });
});
