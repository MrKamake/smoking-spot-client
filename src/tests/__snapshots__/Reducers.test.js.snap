// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Map /> matches snapshot 1`] = `
<Fragment>
  <div
    className="header-box"
  >
    <Link
      className="new-page-button"
      to="/spots/new"
    >
      <img
        alt="new-button"
        className="new-button-icon"
        src="https://www.iconsdb.com/icons/preview/white/plus-5-xxl.png"
      />
    </Link>
    <Link
      className="sub-logo"
      to="/home"
    >
      <div
        className="sub-logo-smoking"
      >
        SMOKING
      </div>
      <div
        className="sub-logo-spot"
      >
        SPOT
      </div>
    </Link>
    <button
      className="search-button"
      onClick={[Function]}
    >
      <img
        alt="search box"
        className="search-button-icon"
        src="https://www.iconsdb.com/icons/preview/white/search-12-xxl.png"
      />
    </button>
  </div>
  <div
    className="map-box"
  >
    <withGoogleMap(Component)
      containerElement={
        <div
          style={
            Object {
              "height": "715px",
              "width": "375px",
            }
          }
        />
      }
      mapElement={
        <div
          style={
            Object {
              "height": "100%",
            }
          }
        />
      }
    />
  </div>
</Fragment>
`;
